File: github_cpp/5/79.cpp
#include <string>
#include "insertion_sort.h"

using namespace std;

InsertionSort::~InsertionSort()
{
    printf("~InsertionSort %s %p\n", Name().c_str(), this);
}

string InsertionSort::Name()
{
    return "InsertionSort";
}

void InsertionSort::Sort(int a[], int n)
{
    for (int i = 1; i < n; i++) {

        
        int b = a[i];

        
        int j = i - 1;

        
        while (b < a[j] && j >= 0) {
            a[j + 1] = a[j];
            j--;
        }

        
        a[j + 1] = b;
    }
}

File: github_java/4/79.java
package Sorting;

public class Insertionsort {
	public static void main(String[] args) {

		int[] arr = { 100, 10, 20, 1, 63, 1, 5 };
		printa(arr, arr.length);
		insertion(arr, arr.length);
	}

	public static void printa(int[] arr, int n) {
		for (int i = 0; i < n; i++)
			System.out.print(arr[i] + " ");

		System.out.println();
	}

	public static void insertion(int[] arr, int n) {
		System.out.println("insertion sort");
		for (int j = 1; j < n; j++) {
			int min = arr[j];
			for (int i = j - 1; i >= 0; i--) {
				if (arr[i] > min) {
					arr[i + 1] = arr[i];
					arr[i] = min;

				}
			}

		}
		printa(arr, n);
	}

}
